libparsec/crates/client/src/workspace/store/cache.rs

```diff
@@ -0,0 +1,411 @@
+                    manifest_id
+                }
+                ArcLocalChildManifest::Folder(m) => m.base.id,
+            };
+            self.manifests.insert(manifest_id, manifest);
+        }
+
+        /// If the update lock is not held, we can still have to insert the manifest in
+        /// cache when populating it.
+        /// Hence this `insert_if_missing` method that handles the case of a populate
+        /// already done by a concurrent operation, in which case the provided manifest
+        /// is simply discarded.
+        ///
+        /// This method returns the manifest in cache.
+        pub fn insert_if_missing(
+            &mut self,
+            manifest: ArcLocalChildManifest,
+        ) -> ArcLocalChildManifest {
+            let manifest_id = match &manifest {
+                ArcLocalChildManifest::File(m) => {
+                    let manifest_id = m.base.id;
+                    assert!(
+                        manifest_id != self.root_manifest_id,
+                        "Root manifest must be a folder !"
+                    );
+                    manifest_id
+                }
+                ArcLocalChildManifest::Folder(m) => m.base.id,
+            };
+
+            match self.manifests.entry(manifest_id) {
+                std::collections::hash_map::Entry::Vacant(entry) => {
+                    entry.insert(manifest.clone());
+                    manifest
+                }
+                // Plot twist: a concurrent operation has updated the cache !
+                // So we discard the data we've fetched and pretend we got a cache hit in
+                // the first place.
+                std::collections::hash_map::Entry::Occupied(entry) => entry.get().to_owned(),
+            }
+        }
+
+        pub fn root_manifest(&self) -> &Arc<LocalFolderManifest> {
+            match self
+                .manifests
+                .get(&self.root_manifest_id)
+                .expect("always present")
+            {
+                ArcLocalChildManifest::Folder(root_manifest) => root_manifest,
+                ArcLocalChildManifest::File(_) => unreachable!("Root manifest must be a folder !"),
+            }
+        }
+    }
+
+    impl std::ops::Deref for ManifestsHashMap {
+        type Target = HashMap<VlobID, ArcLocalChildManifest>;
+        fn deref(&self) -> &Self::Target {
+            &self.manifests
+        }
+    }
+}
+pub(super) use manifest_hash_map::ManifestsHashMap;
+
+#[derive(Debug)]
+pub(super) struct CurrentViewCache {
+    /// `manifests` contains a cache on the database:
+    /// - the cache may be cleared at any given time (i.e. inserting an entry in the cache
+    ///   doesn't guarantee it will be available later on).
+    /// - if the cache is present, it always correspond to the latest value (so the cache
+    ///   should always be preferred over data coming from the database).
+    /// - the root manifest is guaranteed to be always present (even after a clear !).
```

Comment:
It's just an idea and I don't have a clear idea of how it's used yet, but we could keep the root manifest in a separated field, hence avoiding some special cases.